File: frontend/src/api/projects/finance.ts
import { apiFetch } from '../client';

export type ProjectDirectExpense = {
  id: number;
  project_id: number;
  expense_date: string;
  category: string;
  description: string | null;
  amount: number;
  vendor: string | null;
  reference_no: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type ProjectDirectExpenseCreate = {
  expense_date: string;
  category: string;
  description?: string | null;
  amount: number;
  vendor?: string | null;
  reference_no?: string | null;
  notes?: string | null;
};

export type ProjectRevenue = {
  id: number;
  project_id: number;
  revenue_date: string;
  category: string;
  description: string | null;
  amount: number;
  client: string | null;
  reference_no: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type ProjectRevenueCreate = {
  revenue_date: string;
  category: string;
  description?: string | null;
  amount: number;
  client?: string | null;
  reference_no?: string | null;
  notes?: string | null;
};

export type ProjectFinanceListQuery = {
  date_from?: string;
  date_to?: string;
  limit?: number;
  offset?: number;
};

function addCommonParams(params: URLSearchParams, q: ProjectFinanceListQuery) {
  if (q.date_from) params.set('date_from', q.date_from);
  if (q.date_to) params.set('date_to', q.date_to);
  if (q.limit != null) params.set('limit', String(q.limit));
  if (q.offset != null) params.set('offset', String(q.offset));
}

export function createProjectDirectExpense(
  projectId: number,
  payload: ProjectDirectExpenseCreate,
): Promise<ProjectDirectExpense> {
  return apiFetch<ProjectDirectExpense>(`/api/v1/projects/${projectId}/direct-expenses`, {
    method: 'POST',
    body: JSON.stringify({
      ...payload,
      description: payload.description ?? null,
      vendor: payload.vendor ?? null,
      reference_no: payload.reference_no ?? null,
      notes: payload.notes ?? null,
    }),
  });
}

export function listProjectDirectExpenses(
  projectId: number,
  q: ProjectFinanceListQuery,
): Promise<ProjectDirectExpense[]> {
  const params = new URLSearchParams();
  addCommonParams(params, q);
  const qs = params.toString();
  return apiFetch<ProjectDirectExpense[]>(
    `/api/v1/projects/${projectId}/direct-expenses${qs ? `?${qs}` : ''}`,
  );
}

export function createProjectRevenue(
  projectId: number,
  payload: ProjectRevenueCreate,
): Promise<ProjectRevenue> {
  return apiFetch<ProjectRevenue>(`/api/v1/projects/${projectId}/revenues`, {
    method: 'POST',
    body: JSON.stringify({
      ...payload,
      description: payload.description ?? null,
      client: payload.client ?? null,
      reference_no: payload.reference_no ?? null,
      notes: payload.notes ?? null,
    }),
  });
}

export function listProjectRevenues(
  projectId: number,
  q: ProjectFinanceListQuery,
): Promise<ProjectRevenue[]> {
  const params = new URLSearchParams();
  addCommonParams(params, q);
  const qs = params.toString();
  return apiFetch<ProjectRevenue[]>(`/api/v1/projects/${projectId}/revenues${qs ? `?${qs}` : ''}`);
}

File: backend/app/api/v1/projects/finance.py
from __future__ import annotations

import datetime as dt

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.auth import require_permissions
from app.core.database import get_db
from app.modules.projects_finance.schemas import (
    ProjectDirectExpenseCreate,
    ProjectDirectExpensePublic,
    ProjectFinanceListQuery,
    ProjectRevenueCreate,
    ProjectRevenuePublic,
)
from app.modules.projects_finance.service import (
    create_project_direct_expense,
    create_project_revenue,
    list_project_direct_expenses,
    list_project_revenues,
)

router = APIRouter(tags=["Projects • Finance"])


@router.post(
    "/{project_id}/direct-expenses",
    response_model=ProjectDirectExpensePublic,
    dependencies=[Depends(require_permissions({"projects.finance.write"}))],
)
def projects_create_direct_expense(
    project_id: int,
    payload: ProjectDirectExpenseCreate,
    db: Session = Depends(get_db),
) -> ProjectDirectExpensePublic:
    row = create_project_direct_expense(
        db,
        project_id=project_id,
        expense_date=payload.expense_date,
        category=payload.category,
        description=payload.description,
        amount=payload.amount,
        vendor=payload.vendor,
        reference_no=payload.reference_no,
        notes=payload.notes,
    )
    return ProjectDirectExpensePublic.model_validate(row)


@router.get(
    "/{project_id}/direct-expenses",
    response_model=list[ProjectDirectExpensePublic],
    dependencies=[Depends(require_permissions({"projects.finance.read"}))],
)
def projects_list_direct_expenses(
    project_id: int,
    date_from: dt.date | None = None,
    date_to: dt.date | None = None,
    limit: int = 100,
    offset: int = 0,
    db: Session = Depends(get_db),
) -> list[ProjectDirectExpensePublic]:
    q = ProjectFinanceListQuery(
        date_from=date_from,
        date_to=date_to,
        limit=limit,
        offset=offset,
    )
    rows = list_project_direct_expenses(
        db,
        project_id=project_id,
        date_from=q.date_from,
        date_to=q.date_to,
        limit=q.limit,
        offset=q.offset,
    )
    return [ProjectDirectExpensePublic.model_validate(r) for r in rows]


@router.post(
    "/{project_id}/revenues",
    response_model=ProjectRevenuePublic,
    dependencies=[Depends(require_permissions({"projects.finance.write"}))],
)
def projects_create_revenue(
    project_id: int,
    payload: ProjectRevenueCreate,
    db: Session = Depends(get_db),
) -> ProjectRevenuePublic:
    row = create_project_revenue(
        db,
        project_id=project_id,
        revenue_date=payload.revenue_date,
        category=payload.category,
        description=payload.description,
        amount=payload.amount,
        client=payload.client,
        reference_no=payload.reference_no,
        notes=payload.notes,
    )
    return ProjectRevenuePublic.model_validate(row)


@router.get(
    "/{project_id}/revenues",
    response_model=list[ProjectRevenuePublic],
    dependencies=[Depends(require_permissions({"projects.finance.read"}))],
)
def projects_list_revenues(
    project_id: int,
    date_from: dt.date | None = None,
    date_to: dt.date | None = None,
    limit: int = 100,
    offset: int = 0,
    db: Session = Depends(get_db),
) -> list[ProjectRevenuePublic]:
    q = ProjectFinanceListQuery(
        date_from=date_from,
        date_to=date_to,
        limit=limit,
        offset=offset,
    )
    rows = list_project_revenues(
        db,
        project_id=project_id,
        date_from=q.date_from,
        date_to=q.date_to,
        limit=q.limit,
        offset=q.offset,
    )
    return [ProjectRevenuePublic.model_validate(r) for r in rows]
